src/osp/Active/drawing.h

```diff
@@ -24,67 +24,202 @@
+    using storage_type = basic_storage<osp::active::ActiveEnt, osp::active::EntityToDraw>;
+};
+
+//-----------------------------------------------------------------------------
+
 namespace osp::active
 {
 
+using material_family_t = entt::family<struct material_type>;
+using material_id_t = material_family_t::family_type;
+
+struct MaterialCommon { };
+struct MaterialTerrain { };
+struct MaterialSkybox { };
+
+struct ACompMaterial
+{
+    material_id_t m_type;
+};
+
+//-----------------------------------------------------------------------------
+
 /**
- * A function pointer to a Shader's draw() function
- * @param ActiveEnt - The entity being drawn; used to fetch component data
- * @param ActiveScene - The scene containing the entity's component data
- * @param ACompCamera - Camera used to draw the scene
+ * @brief Stores a draw function and user data needed to draw a single entity
+ *
+ * This is usually generated by a DrawAssigner_t
  */
-using ShaderDrawFnc_t = void (*)(
-        ActiveEnt,
-        ActiveScene&,
-        ACompCamera const&) noexcept;
+struct EntityToDraw
+{
+    /**
+     * @brief A function pointer to a Shader's draw() function
+     *
+     * @param ActiveEnt    [in] The entity being drawn; used to fetch component data
+     * @param ActiveScene& [ref] The scene containing the entity's component data
+     * @param ACompCamera  [in] Camera used to draw the scene
+     * @param void*        [in] User data
+     */
+    using ShaderDrawFnc_t = void (*)(
+            ActiveEnt, ActiveScene&, ACompCamera const&, void*) noexcept;
+
+    ShaderDrawFnc_t m_draw;
+    // Non-owning user data passed to draw function, such as the shader
+    void *m_data;
+
+}; // struct EntityToDraw
 
 /**
- * Represents an entity which requests the scene to be drawn to m_target
+ * @brief Tracks a set of entities and their assigned drawing functions
+ *
+ * Each new entity added (in ACtxRenderGroups::m_added) is passed to one of
+ * every RenderGroup's DrawAssigner_t depending on its material type, which is
+ * responsible for setting its draw function and shader.
+ *
+ * DrawAssigners allow each RenderGroup to assign their own shaders for each
+ * kind of material. RenderGroups are intended to be associated with certain
+ * rendering techniques like Forward, Deferred, and Shadow mapping.
+ *
+ * This also works with game-specific modes like Thermal Imaging.
+ */
+struct RenderGroup
+{
+    using Storage_t = entt::storage_traits<ActiveEnt, EntityToDraw>::storage_type;
+    using ArrayView_t = Corrade::Containers::ArrayView<ActiveEnt>;
+
+    /**
+     * @brief Assigns drawing functions and user data (like shaders) to entities
+     *        of a certain material type for each RenderGroup
+     *
+     * @param ActiveScene& [ref] Scene the entities are part of
+     * @param Storage_t&   [ref] Storage to add accepted entities and data to
+     * @param ArrayView_t  [in] Entities to consider
+     */
+    using DrawAssigner_t = std::function<
```

Comment:
needs to be std::function? or can be function-pointer?

Reply:
Draw assigners rely on lambda captures. They're only called once per frame at max, so I don't believe any overhead will be an issue here.